katcp/server.py

```diff
@@ -2318,6 +2338,99 @@ def inform_callback(sensor, reading):
+        min_args = 1
+        max_args = 1 + max(SampleStrategy.SAMPLING_NUM_PARAMS_LOOKUP.values())
+        for num_strategy_args in range(min_args, max_args + 1):
+            args = msg.arguments[-num_strategy_args:]
+            try:
+                strategy = args[0]
+                params = args[1:]
+                if self.PROTOCOL_INFO.strategy_allowed(strategy):
+                    if katcp_version < SEC_TS_KATCP_MAJOR and strategy == 'period':
+                        # Slightly nasty hack, but since period is the only v4 strategy
+                        # involving timestamps it's not _too_ nasty :)
+                        params = [float(params[0]) * MS_TO_SEC_FAC] + params[1:]
+                    requested_strategy = SampleStrategy.get_strategy(
+                        strategy, inform_callback, sensor, *params, ioloop=self.ioloop)
+                    if requested_strategy:
+                        break
+            except Exception:
+                pass
+
+        if not requested_strategy:
+            raise FailReply("No valid strategy provided: %s." % msg.arguments)
+
+        # with known strategy, remaining arguments are sensor names
+        names = msg.arguments[0:-num_strategy_args]
+        invalid_names = []
+        for name in names:
+            if name not in self._sensors:
+                invalid_names.append(name)
+        if invalid_names:
+            raise FailReply("Unknown sensor names: %s." % " ".join(invalid_names))
+
+        for name in names:
+            sensor = self._sensors[name]
+
+            # attempt to set sampling strategy
+            new_strategy = SampleStrategy.get_strategy(
+                strategy, inform_callback, sensor, *params, ioloop=self.ioloop)
+
+            # Remove and cancel old strategy
+            old_strategy = self._strategies[client].pop(sensor, None)
+            if old_strategy:
+                old_strategy.cancel()
+
+            # todo: replace isinstance check with something better
+            if not isinstance(new_strategy, SampleNone):
+                self._strategies[client][sensor] = new_strategy
+                new_strategy.start()
+
+        strategy, params = requested_strategy.get_sampling_formatted()
+        if katcp_version < SEC_TS_KATCP_MAJOR and strategy == 'period':
+            # Another slightly nasty hack, but since period is the only
+            # v4 strategy involving timestamps it's not _too_ nasty :)
+            params = [int(float(params[0]) * SEC_TO_MS_FAC)] + params[1:]
+
+        # Let the ioloop run so that the #sensor-status informs are sent before
+        # the reply. Not strictly neccesary, but a number of tests depend on
+        # this behaviour, less effort to fix it here :-/
+        yield gen.moment
+        reply_args = names + [strategy] + params
+        raise gen.Return(req.make_reply("ok", *reply_args))
+
     @request()
     @return_reply()
     def request_sensor_sampling_clear(self, req):
```

Comment:
Not sure if there is enough of a use case to also update sensor_sampling_clear to support bulk sensors, but that can be done in subsequent PR if required.